commit 33e0b7fd58ec8ce9fd23d3a66a91e9b4b7c2b928
Author: Oleksiy Chernyavskyy <ochern@ics.com>
Date:   Fri Jun 10 23:15:00 2016 +0300

    bug 1654 fix

diff --git a/lib/Xm/LabelG.c b/lib/Xm/LabelG.c
index b674041..726b095 100644
--- a/lib/Xm/LabelG.c
+++ b/lib/Xm/LabelG.c
@@ -35,6 +35,7 @@ static char rcsid[] = "$TOG: LabelG.c /main/24 1999/01/26 15:31:18 mgreess $"
 #include <config.h>
 #endif
 #define FIX_1517
+#define FIX_1654
 
 #include <string.h>
 #include <stdio.h>
@@ -2352,12 +2353,20 @@ LRectangle *background_box)
     if (LabG_StringRect(lw).width < availW - marginal_width)
     	width = LabG_StringRect(lw).width;
     else
+#ifdef FIX_1654
+    	width = availW - marginal_width;
+#else
     	width = availW - marginal_width - x;
+#endif
 
     if (LabG_StringRect(lw).height < availH - marginal_height)
     	height = LabG_StringRect(lw).height;
     else
+#ifdef FIX_1654
+    	height = availH - marginal_height;
+#else
     	height = availH - marginal_height - y;
+#endif
     
     XFillRectangle(XtDisplay(lw), XtWindow(lw), LabG_BackgroundGC(lw),
 		x, y, width, height);
