--- openmotif-2.3.1-old/clients/mwm/WmWinConf.c	2002-01-11 02:25:38.000000000 +0530
+++ openmotif-2.3.1-old/clients/mwm/WmWinConf.c	2013-08-23 12:47:45.000000000 +0530
@@ -1512,7 +1512,7 @@ DrawSegments (Display *dpy, Window win, 
  *************************************<->***********************************/
 void MoveOutline (int x, int y, unsigned int width, unsigned int height)
 {
-    if (wmGD.freezeOnConfig)
+    if (wmGD.freezeOnConfig || !wmGD.pActiveSD->moveOpaque)
     {
 	DrawOutline (x, y, width, height);
     }
